### PR TITLE
initial AutoSAR shim layer

### PR DESCRIPTION
**Pull request by JacobBarthelmeh**
Rebase and squash of old work done (https://github.com/wolfSSL/wolfssl/pull/2667)

### CODE REVIEW

File: wolfcrypt/src/port/autosar/cryif.c [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+void CryIf_Init(const CryIf_ConfigType* in)
+{
+    (void)in;
+    Crypto_Init(NULL);
+}
+
+
+void CryIf_GetVersionInfo(Std_VersionInfoType* ver)
+{
+    (void)ver;
```
Comment (SparkiDev) [inline wolfcrypt/src/port/autosar/cryif.c:50]: Not returning anything here through ver?

File: wolfssl/wolfcrypt/port/autosar/Csm.h [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    CRYPTO_OPERATIONMODE_SINGLECALL = 0x07
+} WOLFSSL_OMODE_TYPE;
+
+
+typedef enum Crypto_ServiceInfoType {
+    CRYPTO_ENCRYPT = 0x03,
+    CRYPTO_DECRYPT = 0x04,
+    CRYPTO_RANDOMGENERATE = 0x0B,
+
+#if 0
```
Comment (SparkiDev) [inline wolfssl/wolfcrypt/port/autosar/Csm.h:86]: Change from #if 0 to #ifdef CSM_UNSUPPORTED_ALGS

File: wolfssl/wolfcrypt/port/autosar/Csm.h [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    CRYPTO_CERTIFICATEVERIFY = 0x12,
+    CRYPTO_KEYSETVALID = 0x13,
+#endif
+} Crypto_ServiceInfoType;
+
+
+typedef enum Crypto_AlgorithmModeType {
+    CRYPTO_ALGOMODE_NOT_SET = 0x00,
+    CRYPTO_ALGOMODE_CBC = 0x02,
+#if 0
```
Comment (SparkiDev) [inline wolfssl/wolfcrypt/port/autosar/Csm.h:110]: Change to a #ifdef

File: wolfcrypt/src/port/autosar/crypto.c [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+            WOLFSSL_MSG("Bogus input key ID redirection (too large)");
+            ret = -1;
+        }
+        else {
+            i = job->jobRedirectionInfoRef->inputKeyId;
+            *key   = keyStore[i].key;
+            *keySz = keyStore[i].keyLen;
+        }
+    }
+    if (eId == job->jobRedirectionInfoRef->secondaryInputKeyElementId) {
```
Comment (SparkiDev) [inline wolfcrypt/src/port/autosar/crypto.c:115]: Should be 'else if'?

File: wolfcrypt/src/port/autosar/crypto.c [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+        if (activeJobs[i].inUse == 1 && activeJobs[i].jobId == job->jobId) {
+            break;
+        }
+    }
+
+    if (i >= MAX_JOBS) {
+        WOLFSSL_MSG("Error finding AES structure");
+    }
+    else {
+        wc_AesFree(&activeJobs[i].aes);
```
Comment (SparkiDev) [inline wolfcrypt/src/port/autosar/crypto.c:206]: Don't see an wc_AesInit().

File: wolfcrypt/src/port/autosar/crypto.c [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+        if (aes == NULL) {
+            WOLFSSL_MSG("Unable to get AES structure for use");
+            return E_NOT_OK;
+        }
+
+        if (wc_AesSetKey(aes, key, keySz, iv, encrypt) != 0) {
+            WOLFSSL_MSG("Crypto error setting up AES key");
+            return E_NOT_OK;
+        }
+        /* ForceZero(key, keySz); Do not keep raw key in memory */
```
Comment (SparkiDev) [inline wolfcrypt/src/port/autosar/crypto.c:255]: Uncomment or remove.

File: wolfcrypt/src/port/autosar/crypto.c [hunk 7]
```diff
@@ -0,0 +1,457 @@
...
+
+    uint8  *out   = job->jobPrimitiveInputOutput.outputPtr;
+    uint32 *outSz = job->jobPrimitiveInputOutput.outputLengthPtr;
+
+    if (outSz == NULL || out == NULL) {
+        WOLFSSL_MSG("Bad parameter passed into wolfSSL_Crypto_RNG");
+        return E_NOT_OK;
+    }
+
+    ret = wc_InitRng_ex(&rng, NULL, 0);
```
Comment (SparkiDev) [inline wolfcrypt/src/port/autosar/crypto.c:336]: We don't want a global?

File: wolfcrypt/src/port/autosar/csm.c [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+{
+    (void)config;
+    CryIf_Init(NULL);
+}
+
+
+/* not yet implemented */
+void Csm_GetVersionInfo(Std_VersionInfoType* version)
+{
+    (void)version;
```
Comment (SparkiDev) [inline wolfcrypt/src/port/autosar/csm.c:162]: Call CryIf_GetVersionInfo()

File: wolfcrypt/src/port/autosar/example.c [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+    const uint8 msg[] = { /* "Now is the time for all " w/o trailing 0 */
+        0x6e,0x6f,0x77,0x20,0x69,0x73,0x20,0x74,
+        0x68,0x65,0x20,0x74,0x69,0x6d,0x65,0x20
+    };
+    const uint8 verify[] =
+    {
+        0x95,0x94,0x92,0x57,0x5f,0x42,0x81,0x53,
+        0x2c,0xcc,0x9d,0x46,0x77,0xa2,0x33,0xcb
+    };
+    const uint8 key[] = "0123456789abcdef   ";  /* align */
```
Comment (SparkiDev) [inline wolfcrypt/src/port/autosar/example.c:49]: What are we aligning on?

File: wolfcrypt/src/port/autosar/example.c [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+int main(int argc, char* argv[])
+{
+    int ret = 0;
+    (void)argv;
+    (void)argc;
+
+    wolfSSL_Debugging_ON();
+    Csm_Init(NULL);
+
+    ret |= run_test(singleshot_test, "singleshot_test ... ");
```
Comment (SparkiDev) [inline wolfcrypt/src/port/autosar/example.c:422]: Don't like seeing test in an example.
Either this is an example showing different uses or this is a test program.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: wolfcrypt/src/port/autosar/crypto.c (hunk 7 above)
Comment (JacobBarthelmeh) [reply wolfcrypt/src/port/autosar/crypto.c:336]: There's no cleanup call Crypto_Cleanup to match the Crypto_Init, is it ok staying alloc'd after program exit with no wc_FreeRng call?

File: wolfcrypt/src/port/autosar/cryif.c (hunk 1 above)
Comment (JacobBarthelmeh) [reply wolfcrypt/src/port/autosar/cryif.c]: we don't have an autosar vendor/model ID but set the major/minor/patch versions now

File: wolfcrypt/src/port/autosar/cryif.c [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    Crypto_Init(NULL);
+}
+
+
+void CryIf_GetVersionInfo(Std_VersionInfoType* ver)
+{
+    if (ver != NULL) {
+        ver->vendorID = 0; /* no vendor or module ID */
+        ver->moduleID = 0;
+        ver->sw_major_version = 5;
```
Comment (SparkiDev) [inline wolfcrypt/src/port/autosar/cryif.c:53]: Use LIBWOLFSSL_VERSION_HEX to extract the values.

File: wolfcrypt/src/port/autosar/crypto.c [hunk 12]
```diff
@@ -0,0 +1,470 @@
...
+
+#ifdef REDIRECTION_CONFIG
+    /* keys should be set... */
+    if (job->jobRedirectionInfoRef == NULL) {
+        WOLFSSL_MSG("Issue with getting key redirection");
+        wc_UnLockMutex(&crypto_mutex);
+        return -1;
+    }
+
+    // @TODO sanity checks on setup... uint8 redirectionConfig;
```
Comment (SparkiDev) [inline wolfcrypt/src/port/autosar/crypto.c:103]: Don't use '//'.

File: wolfcrypt/src/port/autosar/crypto.c (hunk 7 above)
Comment (SparkiDev) [reply wolfcrypt/src/port/autosar/crypto.c:349]: I think so. There should be a cleanup if they care about program exit.
I was also worried about threading.

File: wolfcrypt/src/port/autosar/csm.c [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+}
+
+
+/* getter function for CSM version info */
+void Csm_GetVersionInfo(Std_VersionInfoType* version)
+{
+    if (version != NULL) {
+        version->vendorID = 0; /* no vendor or module ID */
+        version->moduleID = 0;
+        version->sw_major_version = 5;
```
Comment (SparkiDev) [inline wolfcrypt/src/port/autosar/csm.c:165]: Use LIBWOLFSSL_VERSION_HEX to extract the values.

File: wolfcrypt/src/port/autosar/test.c [hunk 14]
```diff
@@ -0,0 +1,430 @@
...
+    uint8 cipher[BLOCK_SIZE * 3];
+    uint8 plain[BLOCK_SIZE * 3];
+
+    uint32 cipherSz = 0;
+    uint32 plainSz  = 0;
+    const uint8 msg[] = { /* "Now is the time for all " w/o trailing 0 */
+        0x6e,0x6f,0x77,0x20,0x69,0x73,0x20,0x74,
+        0x68,0x65,0x20,0x74,0x69,0x6d,0x65,0x20
+    };
+    const uint8 key[] = "0123456789abcdef   ";  /* align */
```
Comment (SparkiDev) [inline wolfcrypt/src/port/autosar/test.c:124]: What's the comment mean?

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: wolfcrypt/src/port/autosar/test.c (hunk 14 above)
Comment (JacobBarthelmeh) [reply wolfcrypt/src/port/autosar/test.c:124]: left over from copy paste of keys that were used in wolfcrypt/test/test.c

**Comment by SparkiDev** [comment]
retest this please

**Review by SparkiDev — APPROVED** [review]

**Pull request merged**